File: Java/Factorial.java
/* ==================================================
 * Factorial
 * Shikuan Huang
 * 
 * Basic recursive implementation
 * Dynamic programming implementation
 * Tail recursive implementation
 * Iterative implementation
 * 
 * Arrays.fill can be used to fill an array of a primitive type.
 *   Arrays.fill(array, val);
 * ==================================================
 */
import java.util.*;
import java.math.BigInteger;

public class Factorial {
  public static void main(String[] args) {
	Scanner input = new Scanner(System.in);
	System.out.println("Factorial of? ");
	long userInput = input.nextLong();
	BigInteger converted = new BigInteger(String.valueOf(userInput));
	System.out.println("Factorial(" + userInput +") = " + iFactorial(converted));
	input.close();
  }
  
  // Basic recursive implementation
  static BigInteger rFactorial(BigInteger arg) {
	if (arg.equals(BigInteger.ZERO) || arg.equals(BigInteger.ONE)) {
      return BigInteger.ONE;
	}
    return arg.multiply(rFactorial(arg.subtract(BigInteger.ONE)));
  }
  
  // Dynamic programming implementation
  static BigInteger dFactorial(BigInteger arg) {
    BigInteger[] answers = new BigInteger[arg.intValue() + 1];
    Arrays.fill(answers, new BigInteger(String.valueOf(-1)));
    return dFactorialRecursiveHelper(arg, answers);
  }
  
  static BigInteger dFactorialRecursiveHelper(BigInteger arg, BigInteger[] answers) {
	BigInteger NegOne = new BigInteger(String.valueOf(-1));
	if (!answers[arg.intValue()].equals(NegOne)) {
	  return answers[arg.intValue()];
	}

	if (arg.equals(BigInteger.ZERO) || arg.equals(BigInteger.ONE)) {
	  answers[arg.intValue()] = arg;
	  return arg;
	}
	
	answers[arg.intValue()] = 
	    arg.multiply(dFactorialRecursiveHelper(arg.subtract(BigInteger.ONE), answers));
	return answers[arg.intValue()];
  }
  
  // Tail recursive implementation
  static BigInteger tFactorial(BigInteger arg) {
	return tFactorial(arg, BigInteger.ONE);
  }
  
  static BigInteger tFactorial(BigInteger arg, BigInteger partialAns) {
	if (arg.equals(BigInteger.ZERO) || arg.equals(BigInteger.ONE)) {
	  return partialAns;
	}
	
	return tFactorial(arg.subtract(BigInteger.ONE), partialAns.multiply(arg));
  }
  
  // Iterative implementation
  static BigInteger iFactorial(BigInteger arg) {
	BigInteger argCopy = arg;
	BigInteger partialSum = new BigInteger(String.valueOf(1));
	for (int i = argCopy.intValue(); i > 1; i--) {
	  partialSum = partialSum.multiply(argCopy);
	  argCopy = argCopy.subtract(BigInteger.ONE);
	}
	return partialSum;
  }
}

File: CPP/Factorial.cpp
/* ==================================================
Factorial
Shikuan Huang

Basic recursive implementation
Dynamic programming implementation
Tail recursive implementation
Iterative implementation
================================================== */

#include <iostream>
using namespace std;

// Basic recursive implementation
long long rFactorial(long long number) {
	if (number == 0 || number == 1) {
		return 1;
	}
	return number * rFactorial(number - 1);
}

// Dynamic programming implementation
long long dFactorial(long long number, long long * answers) {
	if (answers[number] != -1) {
		return answers[number];
	}
	
	if (number == 0 || number == 1) {
		answers[number] = number;
		return number;
	}
	
	answers[number] = number * dFactorial(number - 1, answers);
	return answers[number];
}

long long dFactorial(long long number) {
	long long * answers = new long long[number + 1];
	for (int i = 0; i < number + 1; i++) {
		answers[i] = -1;
	}
	long long answer = dFactorial(number, answers);
	delete[] answers;
	return answer;
}

// Tail recursive implementation
long long tFactorial(int number, long long total = 1) {
	if (number == 0 || number == 1) {
		return total;
	}
	return tFactorial(number - 1, number * total);
}

// Iterative implementation
long long iFactorial(int number) {
	long long total = 1;
	while (number != 1) {
		total = total * number;
		number--;
	}
	return total;
}

int main() {
	int input;
	
	bool end = false;
	while(!end) {
		cout << "Factorial for? ";
		cin >> input;
		
		if (input == -1) {
			end = true;
		}
		else if (input < 0) {
			cout << "Invalid: " << input << endl;	
		}
		else {
			cout << "Factorial(" << input << ") = " << dFactorial(input) <<endl;
		}
	}
}
